src/Orleans.Runtime/GrainDirectory/GrainLocator.cs

```diff
@@ -0,0 +1,120 @@
+using Orleans.GrainDirectory;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    /// <summary>
+    /// Implementation of <see cref="IGrainLocator"/> that uses an <see cref="IGrainDirectory"/> store.
+    /// </summary>
+    internal class GrainLocator : IGrainLocator
+    {
+        private readonly IGrainDirectory grainDirectory;
+        private readonly InClusterGrainLocator inClusterGrainLocator;
+        private readonly IGrainDirectoryCache cache;
+
+        public GrainLocator(IGrainDirectory grainDirectory, InClusterGrainLocator inClusterGrainLocator)
+        {
+            this.grainDirectory = grainDirectory;
+            this.inClusterGrainLocator = inClusterGrainLocator;
+            this.cache = new LRUBasedGrainDirectoryCache(GrainDirectoryOptions.DEFAULT_CACHE_SIZE, GrainDirectoryOptions.DEFAULT_MAXIMUM_CACHE_TTL);
+        }
+
+        public async Task<List<ActivationAddress>> Lookup(GrainId grainId)
+        {
+            if (grainId.IsClient)
+                return await this.inClusterGrainLocator.Lookup(grainId);
+
+            var results = new List<ActivationAddress>();
+
+            var entry = await this.grainDirectory.Lookup(grainId.ToParsableString());
+
+            if (entry == null)
+                return results;
+
+            results.Add(ConvertToActivationAddress(entry));
+            this.cache.AddOrUpdate(grainId, results.Select(item => Tuple.Create(item.Silo, item.Activation)).ToList(), 0);
+
+            return results;
+        }
+
+        public async Task<ActivationAddress> Register(ActivationAddress address)
+        {
+            if (address.Grain.IsClient)
+                return await this.inClusterGrainLocator.Register(address);
+
+            var result = await this.grainDirectory.Register(ConvertToGrainAddress(address));
+            var activationAddress = ConvertToActivationAddress(result);
+            this.cache.AddOrUpdate(
+                activationAddress.Grain,
+                new List<Tuple<SiloAddress,ActivationId>>() { Tuple.Create(activationAddress.Silo, activationAddress.Activation) },
+                0);
+            return activationAddress;
+        }
+
+        public bool TryLocalLookup(GrainId grainId, out List<ActivationAddress> addresses)
+        {
+            if (grainId.IsClient)
+                return this.inClusterGrainLocator.TryLocalLookup(grainId, out addresses);
+
+            if (this.cache.LookUp(grainId, out var results))
+            {
+                addresses = results
+                    .Select(tuple => ActivationAddress.GetAddress(tuple.Item1, grainId, tuple.Item2))
+                    .ToList();
+                return true;
+            }
+
+            addresses = null;
+            return false;
+        }
+
+        public async Task Unregister(ActivationAddress address, UnregistrationCause cause)
+        {
+            if (address.Grain.IsClient)
+                await this.inClusterGrainLocator.Unregister(address, cause);
```

Comment:
We aren't returning after this, so we're going to try unregistering the client from two directories, even though we only registered it in one.

Might be worthwhile enforcing braces around conditional/loop bodies. It's one of the things which Go enforces even though they are light on syntax in other places, because omitting them can make errors more subtle.

Reply:
Bad copy-paste on my part while trying to split this into multiple PR. Other method just return from this if statement